pkg/scheduler/plugins/gang/cache.go

```diff
@@ -0,0 +1,222 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gang
+
+import (
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clientV1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
+	pgLister "sigs.k8s.io/scheduler-plugins/pkg/generated/listers/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingconfig "github.com/koordinator-sh/koordinator/apis/scheduling/config"
+)
+
+type GangCache struct {
+	lock       *sync.RWMutex
+	gangItems  map[string]*Gang
+	pluginArgs *schedulingconfig.GangArgs
+	podLister  clientV1.PodLister
+	pgLister   pgLister.PodGroupLister
+}
+
+func NewGangCache(args *schedulingconfig.GangArgs, podLister clientV1.PodLister, pgLister pgLister.PodGroupLister) *GangCache {
+	return &GangCache{
+		gangItems:  make(map[string]*Gang),
+		lock:       new(sync.RWMutex),
+		pluginArgs: args,
+		podLister:  podLister,
+		pgLister:   pgLister,
+	}
+}
+
+func (gangCache *GangCache) createGangToCache(gangId string) *Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	if gang, exist := gangCache.gangItems[gangId]; exist {
+		return gang
+	}
+
+	gang := NewGang(gangId)
+	gangCache.gangItems[gangId] = gang
+	return gang
+}
+
+func (gangCache *GangCache) getGangFromCache(gangId string) *Gang {
+	gangCache.lock.RLock()
+	defer gangCache.lock.RUnlock()
+
+	return gangCache.gangItems[gangId]
+}
+
+func (gangCache *GangCache) deleteGangFromCache(gangId string) {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	delete(gangCache.gangItems, gangId)
+}
+func (gangCache *GangCache) getGangItems() map[string]*Gang {
+	gangCache.lock.Lock()
+	defer gangCache.lock.Unlock()
+
+	return gangCache.gangItems
+}
+
+func (gangCache *GangCache) RecoverGangCache() {
+	// recover podGroup First
+	pgLister := gangCache.pgLister
+	pgList, err := pgLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podGroupList List error: %v", err.Error())
+	}
+	for _, podGroup := range pgList {
+		gangCache.onPodGroupAdd(podGroup)
+
+	}
+	// recover gang from pod
+	podLister := gangCache.podLister
+	podsList, err := podLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podsList List error: %v", err.Error())
+	}
+	for _, pod := range podsList {
+		gangCache.onPodAdd(pod)
+
+		if pod.Spec.NodeName != "" {
+			gangName := getGangNameByPod(pod)
+			if gangName == "" {
+				continue
+			}
+			gangId := getNamespaceSplicingName(pod.Namespace,
+				gangName)
+			gang := gangCache.getGangFromCache(gangId)
+			if gang == nil {
+				klog.Errorf("Didn't find gang when recover by pod, pod:%v, gang: %v", pod.Name, gangId)
+				continue
+			}
+			gang.addBoundPod(pod)
+			gang.setResourceSatisfied()
```

Comment:
set gang as resource satisfied status if only a child pod is bound?

Reply:
if a pod is bound, which means the gang has already passed the Permit stage check, so the gang must be  resouceSatisfied